parlai/agents/rag/retrievers.py

```diff
@@ -0,0 +1,1217 @@
+            'poly_dropout_reduction_type': model_opt['poly_dropout_reduction_type'],
+            'poly_dropout_use_codes': model_opt.get('poly_dropout_use_codes', True),
+        }
+        logging.disable()
+        agent = create_agent(create_model_opt)
+        logging.enable()
+        assert isinstance(agent, DropoutPolyAgent)
+        return agent.model, agent.dict
+
+    def forward(self, query_vecs: torch.LongTensor) -> torch.Tensor:
+        if self.use_codes:
+            ctxt_rep, ctxt_mask, _ = self.model(ctxt_tokens=query_vecs)
+        else:
+            model = self.model.module if hasattr(self.model, 'module') else self.model
+            old_type = model.type
+            model.type = 'n_first'  # type: ignore
+
+            ctxt_rep, ctxt_mask, _ = self.model(ctxt_tokens=query_vecs)
+            model.type = old_type  # type: ignore
+
+        ctxt_rep = reduce_ctxt(ctxt_rep, ctxt_mask, self.reduction_type)
+        return ctxt_rep
+
+    @property
+    def embedding_size(self) -> int:
+        model = self.model.module if hasattr(self.model, 'module') else self.model
+        return model.encoder_ctxt.embedding_size  # type: ignore
+
+
+class RagDprEncoder(TransformerEncoder, RagRetrievalEncoder):
+    """
+    Basically provide a wrapper around TransformerEncoder to load RAG Query/Document
+    models.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        dpr_model: str = 'bert',
+        pretrained_path: str = DPR_ZOO_MODEL,
+        encoder_type: str = 'query',
+    ):
+        # Override options
+        config: BertConfig = BertConfig.from_pretrained('bert-base-uncased')
+        pretrained_path = modelzoo_path(opt['datapath'], pretrained_path)
+        if not os.path.exists(pretrained_path):
+            # when initializing from parlai rag models, the pretrained path
+            # may not longer exist. This is fine if we've alread trained
+            # the model.
+            assert dpr_model == 'bert_from_parlai_rag'
+            logging.error(f'Pretrained Path does not exist: {pretrained_path}')
+            pretrained_path = modelzoo_path(opt['datapath'], DPR_ZOO_MODEL)
+            dpr_model = 'bert'
+            logging.error(f'Setting to zoo model: {pretrained_path}')
+        enc_opt = {
+            "n_heads": config.num_attention_heads,
+            "n_layers": config.num_hidden_layers,
+            "embedding_size": config.hidden_size,
+            "ffn_size": config.intermediate_size,
+            "dropout": config.hidden_dropout_prob,
+            "attention_dropout": config.attention_probs_dropout_prob,
+            "activation": config.hidden_act,
+            "variant": 'xlm',
+            "reduction_type": 'first',
+            "n_positions": config.max_position_embeddings,
+            "n_segments": config.type_vocab_size,
+        }
+        embedding = torch.nn.Embedding(
+            config.vocab_size, config.hidden_size, padding_idx=config.pad_token_id
+        )
```

Comment:
I think you can just not pass in an embedding and `TransformerEncoder` will do this for you.

Reply:
actually it raises an assertion error; we require that you create an embedding

Reply:
Oh! Interesting.